File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/common/ui/MakeTransferFragment.java
package org.mifos.mobilewallet.mifospay.common.ui;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomSheetBehavior;
import android.support.design.widget.BottomSheetDialog;
import android.support.design.widget.BottomSheetDialogFragment;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import org.mifos.mobilewallet.core.domain.model.PaymentMethod;
import org.mifos.mobilewallet.mifospay.R;
import org.mifos.mobilewallet.mifospay.base.BaseActivity;
import org.mifos.mobilewallet.mifospay.common.TransferContract;
import org.mifos.mobilewallet.mifospay.common.presenter.MakeTransferPresenter;
import org.mifos.mobilewallet.mifospay.data.local.LocalRepository;
import org.mifos.mobilewallet.mifospay.payments.ui.SendFragment;
import org.mifos.mobilewallet.mifospay.utils.Constants;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by naman on 30/8/17.
 */

public class MakeTransferFragment extends BottomSheetDialogFragment
        implements TransferContract.TransferView {

    @Inject
    MakeTransferPresenter mPresenter;

    @Inject
    LocalRepository localRepository;

    TransferContract.TransferPresenter mTransferPresenter;

    @BindView(R.id.btn_confirm)
    Button btnConfirm;

    @BindView(R.id.btn_cancel)
    Button btnCancel;

    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    @BindView(R.id.tv_amount)
    TextView tvAmount;

    @BindView(R.id.tv_client_name)
    TextView tvClientName;

    @BindView(R.id.tv_client_vpa)
    TextView tvClientVpa;

    @BindView(R.id.tv_transfer_status)
    TextView tvTransferStatus;

    @BindView(R.id.ll_content)
    View contentView;

    @BindView(R.id.view_transfer_success)
    View viewTransferSuccess;

    @BindView(R.id.view_transfer_failure)
    View viewTransferFailure;

    private BottomSheetBehavior mBehavior;
    private long toClientId;
    private double amount;
    private String toClientIdentifier;
    private PaymentMethod paymentMethod;

    public static MakeTransferFragment newInstance(
            String toClientIdentifier, PaymentMethod paymentMethod, double amount) {

        Bundle args = new Bundle();
        args.putString(Constants.TO_CLIENT_IDENTIFIER, toClientIdentifier);
        args.putDouble(Constants.AMOUNT, amount);
        args.putSerializable(Constants.PAYMENT_METHOD, paymentMethod);
        MakeTransferFragment fragment = new MakeTransferFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        BottomSheetDialog dialog = (BottomSheetDialog) super.onCreateDialog(savedInstanceState);

        View view = View.inflate(getContext(), R.layout.fragment_make_transfer, null);

        dialog.setContentView(view);
        mBehavior = BottomSheetBehavior.from((View) view.getParent());

        ButterKnife.bind(this, view);
        mPresenter.attachView(this);

        amount = getArguments().getDouble(Constants.AMOUNT);
        paymentMethod = (PaymentMethod) getArguments().getSerializable(Constants.PAYMENT_METHOD);
        toClientIdentifier = getArguments().getString(Constants.TO_CLIENT_IDENTIFIER);
        if (paymentMethod == PaymentMethod.VPA) {
            mTransferPresenter.fetchClient(toClientIdentifier);
        } else {
            showConfirmMsisdnScreen(toClientIdentifier);
        }

        btnCancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });

        btnConfirm.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                tvTransferStatus.setText(Constants.SENDING_MONEY);
                progressBar.setVisibility(View.VISIBLE);
                contentView.setVisibility(View.GONE);
                if (paymentMethod == PaymentMethod.VPA) {
                    mTransferPresenter.fetchToClientAccount(toClientId, amount);
                } else {
                    mTransferPresenter.makeTransferUsingMsisdn(toClientIdentifier, amount);
                }
            }
        });

        return dialog;
    }

    private void showConfirmMsisdnScreen(String msisdnNumber) {
        tvClientName.setText(msisdnNumber);
        tvAmount.setText(Constants.RUPEE + " " + amount);

        contentView.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showToClientDetails(long clientId, String name, String clientIdentifier) {
        this.toClientId = clientId;

        tvClientName.setText(name);
        tvAmount.setText(Constants.RUPEE + " " + amount);
        tvClientVpa.setText(clientIdentifier);

        contentView.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.GONE);

    }

    @Override
    public void transferSuccess() {
        tvTransferStatus.setText(Constants.TRANSACTION_SUCCESSFUL);
        progressBar.setVisibility(View.GONE);
        viewTransferSuccess.setVisibility(View.VISIBLE);
    }

    @Override
    public void transferFailure() {
        tvTransferStatus.setText(Constants.UNABLE_TO_PROCESS_TRANSFER);
        progressBar.setVisibility(View.GONE);
        viewTransferFailure.setVisibility(View.VISIBLE);
    }

    @Override
    public void onStart() {
        super.onStart();
        mBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
    }

    @Override
    public void setPresenter(TransferContract.TransferPresenter presenter) {
        this.mTransferPresenter = presenter;
    }

    @Override
    public void showVpaNotFoundSnackbar() {
        if (getTargetFragment() != null && paymentMethod == PaymentMethod.VPA) {
            getTargetFragment().onActivityResult(SendFragment.REQUEST_SHOW_DETAILS,
                    Activity.RESULT_CANCELED, null);
            dismiss();
        }
    }
}

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/common/ui/MakeTransferFragment.kt
package org.mifos.mobilewallet.mifospay.common.ui

import android.app.Activity
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetBehavior.BottomSheetCallback
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import androidx.transition.TransitionManager
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobilewallet.mifospay.R
import org.mifos.mobilewallet.mifospay.base.BaseActivity
import org.mifos.mobilewallet.mifospay.common.TransferContract
import org.mifos.mobilewallet.mifospay.common.presenter.MakeTransferPresenter
import org.mifos.mobilewallet.mifospay.data.local.LocalRepository
import org.mifos.mobilewallet.mifospay.payments.ui.SendFragment
import org.mifos.mobilewallet.mifospay.utils.Constants
import org.mifos.mobilewallet.mifospay.utils.Toaster
import javax.inject.Inject

/**
 * Created by naman on 30/8/17.
 */
@AndroidEntryPoint
class MakeTransferFragment : BottomSheetDialogFragment(), TransferContract.TransferView {
    @JvmField
    @Inject
    var mPresenter: MakeTransferPresenter? = null

    @JvmField
    @Inject
    var localRepository: LocalRepository? = null
    var mTransferPresenter: TransferContract.TransferPresenter? = null

    @JvmField
    @BindView(R.id.ll_make_transfer_container)
    var makeTransferContainer: ViewGroup? = null

    @JvmField
    @BindView(R.id.btn_confirm)
    var btnConfirm: Button? = null

    @JvmField
    @BindView(R.id.btn_cancel)
    var btnCancel: Button? = null

    @JvmField
    @BindView(R.id.progressBar)
    var progressBar: ProgressBar? = null

    @JvmField
    @BindView(R.id.tv_amount)
    var tvAmount: TextView? = null

    @JvmField
    @BindView(R.id.tv_client_name)
    var tvClientName: TextView? = null

    @JvmField
    @BindView(R.id.tv_client_vpa)
    var tvClientVpa: TextView? = null

    @JvmField
    @BindView(R.id.tv_transfer_status)
    var tvTransferStatus: TextView? = null

    @JvmField
    @BindView(R.id.ll_content)
    var contentView: View? = null

    @JvmField
    @BindView(R.id.view_transfer_success)
    var viewTransferSuccess: View? = null

    @JvmField
    @BindView(R.id.view_transfer_failure)
    var viewTransferFailure: View? = null
    private var mBehavior: BottomSheetBehavior<*>? = null
    private var toClientId: Long = 0
    private var amount = 0.0

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState) as BottomSheetDialog
        val view = View.inflate(context, R.layout.fragment_make_transfer, null)
        dialog.setContentView(view)
        mBehavior = BottomSheetBehavior.from(view.parent as View)
        ButterKnife.bind(this, view)
        mPresenter?.attachView(this)
        amount = arguments?.getDouble(Constants.AMOUNT) ?: 0.0
        mTransferPresenter?.fetchClient(arguments?.getString(Constants.TO_EXTERNAL_ID))
        btnCancel?.setOnClickListener { dismiss() }
        btnConfirm?.setOnClickListener {
            mTransferPresenter?.makeTransfer(
                localRepository?.clientDetails?.clientId ?: 0,
                toClientId, amount
            )
            makeTransferContainer?.let { it1 ->
                TransitionManager.beginDelayedTransition(
                    it1
                )
            }
            tvTransferStatus?.text = Constants.SENDING_MONEY
            progressBar?.visibility = View.VISIBLE
            contentView?.visibility = View.GONE
        }
        return dialog
    }

    override fun showToClientDetails(clientId: Long, name: String?, externalId: String?) {
        toClientId = clientId
        makeTransferContainer?.let { TransitionManager.beginDelayedTransition(it) }
        tvClientName?.text = name
        tvAmount?.text = Constants.RUPEE + " " + amount
        tvClientVpa?.text = externalId
        contentView?.visibility = View.VISIBLE
        progressBar?.visibility = View.GONE
    }

    override fun transferSuccess() {
        tvTransferStatus?.text = Constants.TRANSACTION_SUCCESSFUL
        progressBar?.visibility = View.GONE
        viewTransferSuccess?.visibility = View.VISIBLE
    }

    override fun transferFailure() {
        makeTransferContainer?.let { TransitionManager.beginDelayedTransition(it) }
        tvTransferStatus?.text = Constants.UNABLE_TO_PROCESS_TRANSFER
        progressBar?.visibility = View.GONE
        viewTransferFailure?.visibility = View.VISIBLE
    }

    override fun onStart() {
        super.onStart()
        mBehavior?.state = BottomSheetBehavior.STATE_EXPANDED
    }

    override fun setPresenter(presenter: TransferContract.TransferPresenter?) {
        mTransferPresenter = presenter
    }

    override fun showVpaNotFoundSnackbar() {
        if (targetFragment != null) {
            targetFragment?.onActivityResult(
                SendFragment.REQUEST_SHOW_DETAILS,
                Activity.RESULT_CANCELED, null
            )
            dismiss()
        } else {
            Toaster.showToast(context, "Invalid")
            dismiss()
        }
    }

    override fun enableDragging(enable: Boolean) {
        mBehavior?.setBottomSheetCallback(object : BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                if (newState == BottomSheetBehavior.STATE_DRAGGING && !enable) {
                    mBehavior?.state = BottomSheetBehavior.STATE_EXPANDED
                }
            }

            override fun onSlide(bottomSheet: View, slideOffset: Float) {}
        })
    }

    companion object {
        @JvmStatic
        fun newInstance(toExternalId: String?, amount: Double): MakeTransferFragment {
            val args = Bundle()
            args.putString(Constants.TO_EXTERNAL_ID, toExternalId)
            args.putDouble(Constants.AMOUNT, amount)
            val fragment = MakeTransferFragment()
            fragment.arguments = args
            return fragment
        }
    }
}